langchain4j/src/main/java/dev/langchain4j/service/WithSources.java

```diff
@@ -0,0 +1,17 @@
+package dev.langchain4j.service;
+
+import dev.langchain4j.data.message.AugmentedMessage;
+import lombok.Builder;
+import lombok.Getter;
+
+/**
+ * Represents a container holding augmented information associated with a response.
+ *
+ * @param <T> The type of the response.
+ */
+@Getter
+@Builder
+public class WithSources<T> {
```

Comment:
After a second thought, I think we should make this class a bit more generic, to accomodate for other useful information, such as:
- tools used during the call 
- token usage
- etc

(Please do not add these in this PR.)

Any ideas for the good name for the class that can hold that?

Reply:
`ResponseMessage` or `ResponsePrompt`

Well when we talk about framework like Spring or the javax package, they provide powerful tools for handling HTTP requests and responses. One key concept is how they handle responses.

In Spring, for example, ResponseEntity is a class that encapsulates an HTTP response, including the response body and additional metadata such as headers and status codes. It's a neat way to package up all the necessary information about a response in one object.

Similarly, in the javax package, there's the ResponseBuilder class. This class allows you to construct an HTTP response step by step, setting things like the status code, headers, and entity (the response body). It's a flexible and convenient way to craft custom responses 

Now, let's bring this analogy back to our prompt usecase. Just as ResponseEntity and ResponseBuilder provide a structured way to handle HTTP responses. ResponseMessage class aims to provide a structured way to handle prompts response

By using ResponseMessage (or ResponsePrompt), we can create responses that not only contain the text of the prompt itself but also include additional contextual information, just like how ResponseEntity includes metadata with the response body. This makes our response handling more organized, easier to manage, and more adaptable to different scenarios
```

ResponseMessage responseMessage = ResponseMessage.create("How can I assist you today?")
    .contents(...)
    .tools(...)
    .tokens(...)
    .build();
```

I can include this in a seperated PR if needed

Reply:
I am not sure about these names, they are not really responses, and not messages/prompts.

These object can contain much more:
- Answer/Response itself (`String` or `CustomPOJO` or smth else)
- Retrieval:
  - `List<Content> retrievedContents`
  - Used queries (not sure if needed)
  - Used retrievers (not sure if needed)
- LLM calls:
  - `TokenUsage` (for all calls combined and/or for each call separately)
  - `FinishReason` (of the last call)
  - Model parameters: `modelName`/`temperature`/etc?
  - List of all calls with requests and responses (not sure if needed)?
- Tool executions:
  - Executed tools and their inputs/outputs

I am having a hard time to find a good name right now, maybe it is an indicator of a bad design, or maybe I need to go and rest a bit 😆 

Have a nice weekend!